tests/unit/remote/test_s3.py

```diff
@@ -0,0 +1,91 @@
+import boto3
+import pytest
+from moto import mock_s3
+
+from dvc.remote.s3 import RemoteS3
+
+
+@pytest.fixture
+def s3():
+    """Returns a connection to a bucket with the following file structure:
+
+        bucket
+        ├── data
+        │  ├── alice
+        │  ├── alpha
+        │  └── subdir
+        │     ├── 1
+        │     ├── 2
+        │     └── 3
+        ├── empty_dir
+        ├── empty_file
+        └── foo
+    """
+    with mock_s3():
+        s3 = boto3.client("s3", region_name="us-east-1")
+        s3.create_bucket(Bucket="bucket")
+
+        s3.put_object(Bucket="bucket", Key="empty_dir/")
+        s3.put_object(Bucket="bucket", Key="empty_file", Body=b"")
+        s3.put_object(Bucket="bucket", Key="foo", Body=b"foo")
+        s3.put_object(Bucket="bucket", Key="data/alice", Body=b"alice")
+        s3.put_object(Bucket="bucket", Key="data/alpha", Body=b"alpha")
+        s3.put_object(Bucket="bucket", Key="data/subdir/1", Body=b"1")
+        s3.put_object(Bucket="bucket", Key="data/subdir/2", Body=b"2")
+        s3.put_object(Bucket="bucket", Key="data/subdir/3", Body=b"3")
+
+        yield s3
+
+
+@pytest.fixture
+def remote():
+    """Returns RemoteS3 instance to work with the `s3` fixture."""
+    yield RemoteS3(None, {"url": "s3://bucket", "region": "us-east-1"})
```

Comment:
I think I know where that creds confusion was coming from.  This `RemoteS3` will be calling boto3.client() once again and this fixture is detached from s3 fixture. You don't really need s3 fixture separately, and if you combine them like so:
```
@pytest.fixture                                                               
def remote():                                                                 
    with mock_s3():                                                           
        remote = RemoteS3(None, {"url": "s3://bucket"})                       
        remote.s3.create_bucket(Bucket="bucket")                              
        remote.s3.put_object(Bucket="bucket", Key="empty_dir/")               
        remote.s3.put_object(Bucket="bucket", Key="empty_file", Body=b"")     
        remote.s3.put_object(Bucket="bucket", Key="foo", Body=b"foo")         
        remote.s3.put_object(Bucket="bucket", Key="data/alice", Body=b"alice")
        remote.s3.put_object(Bucket="bucket", Key="data/alpha", Body=b"alpha")
        remote.s3.put_object(Bucket="bucket", Key="data/subdir/1", Body=b"1") 
        remote.s3.put_object(Bucket="bucket", Key="data/subdir/2", Body=b"2") 
        remote.s3.put_object(Bucket="bucket", Key="data/subdir/3", Body=b"3") 
        yield remote                                                          
```
you will be able to ensure that this is the same s3 instance being used both when we are creating stuff and when we are using it in the Remote. And also that way you don't need to ensure that you are passing the same config options to boto3.client() and RemoteS3(), since they will be passed only once to RemoteS3() and automatically passed to boto3.client() inside RemoteS3.s3.

Reply:
looks good, @efiop , I'll change it